sycl/test-e2e/CompositeDevice/composite_device.cpp

```diff
@@ -0,0 +1,228 @@
+  //      APIs may be called even when using other backends, but they will
+  //      return an empty list of composite devices.
+  //   D. The execution environment for a SYCL application has a fixed number of
+  //      composite devices which does not vary as the application executes. As
+  //      a result, each call to these functions returns the same set of device
+  //      objects, and the order of those objects does not vary between calls.
+  {
+    std::vector<sycl::device> AllCompositeDevs = get_composite_devices();
+    std::vector<sycl::device> CombinedCompositeDevs;
+    for (const auto &P : Platforms) {
+      auto CompositeDevs = P.ext_oneapi_get_composite_devices();
+      bool IsL0 = isL0Backend(P.get_backend());
+      // Check C.
+      if (!IsL0 || !IsCombined)
+        assert(CompositeDevs.empty());
+
+      for (const auto &D : CompositeDevs) {
+        if (std::find(CombinedCompositeDevs.begin(),
+                      CombinedCompositeDevs.end(),
+                      D) == CombinedCompositeDevs.end())
+          CombinedCompositeDevs.push_back(D);
+      }
+    }
+    // Check A. and B.
+    assert(AllCompositeDevs.size() == CombinedCompositeDevs.size());
+    for (size_t i = 0; i < AllCompositeDevs.size(); ++i) {
+      const auto &D1 = AllCompositeDevs[i];
+      const auto &D2 = CombinedCompositeDevs[i];
+      assert(D1 == D2);
+      assert(D1.has(sycl::aspect::ext_oneapi_is_composite));
+      assert(D2.has(sycl::aspect::ext_oneapi_is_composite));
+    }
+
+    // Check D.
+    std::vector<sycl::device> AllCompositeDevs2 = get_composite_devices();
+    std::vector<sycl::device> CombinedCompositeDevs2;
+    for (const auto &P : Platforms) {
+      auto CompositeDevs = P.ext_oneapi_get_composite_devices();
+      bool IsL0 = isL0Backend(P.get_backend());
+      // Check C.
+      if (!IsL0 || !IsCombined)
+        assert(CompositeDevs.empty());
+
+      for (const auto &D : CompositeDevs) {
+        if (std::find(CombinedCompositeDevs2.begin(),
+                      CombinedCompositeDevs2.end(),
+                      D) == CombinedCompositeDevs2.end())
+          CombinedCompositeDevs2.push_back(D);
+      }
+    }
+    assert(AllCompositeDevs.size() == AllCompositeDevs2.size());
+    assert(CombinedCompositeDevs.size() == CombinedCompositeDevs2.size());
+    for (size_t i = 0; i < AllCompositeDevs.size(); ++i) {
+      assert(AllCompositeDevs[i] == AllCompositeDevs2[i]);
+      assert(CombinedCompositeDevs[i] == CombinedCompositeDevs2[i]);
+    }
+  }
+
+  // Check that device::info::component_devices:
+  //   A. Returns the set of component devices that are contained by a composite
+  //      device (at least 2).
+  //   B. If "this" device is not a composite device, returns an empty vector.
+  {
+    auto Devs = sycl::device::get_devices();
+    for (const auto &D : Devs) {
+      bool IsL0 = isL0Backend(D.get_backend());
+      if (!IsL0 || !IsCombined)
+        continue;
```

Comment:
I'm very surprised to see this check for `IsL0` in a test (it was just a performance optimization in the implementation)